File: TDDMicroExercises.YoursSolutions/CSharp-benwu/TurnTicketDispenser.Tests/TicketDispenserTest.cs
using NUnit.Framework;

namespace TDDMicroExercises.TurnTicketDispenser.Tests
{
    [TestFixture]
    public class TicketDispenserTest
    {
        [Test]
        public void GIVEN_dispense_a_ticket_WHEN_dispense_a_new_ticket_at_the_same_machine_THEN_turn_number_of_the_new_ticket_is_greater()
        {
            TicketDispenser ticketDispenser = new TicketDispenser();
            TurnTicket oldTicket = ticketDispenser.GetTurnTicket();
            TurnTicket newTicket = ticketDispenser.GetTurnTicket();
            Assert.Greater(newTicket.TurnNumber, oldTicket.TurnNumber);
        }

        [Test]
        public void GIVEN_dispense_a_ticket_WHEN_dispense_a_new_ticket_at_another_machine_THEN_turn_number_of_the_new_ticket_is_greater()
        {
            TicketDispenser ticketDispenser = new TicketDispenser();
            TurnTicket oldTicket = ticketDispenser.GetTurnTicket();

            TicketDispenser anotherTicketDispenser = new TicketDispenser();
            TurnTicket newTicket = anotherTicketDispenser.GetTurnTicket();

            Assert.Greater(newTicket.TurnNumber, oldTicket.TurnNumber);
        }

        [Test]
        public void GIVEN_dispense_ticket_number_10_WHEN_dispense_a_new_ticket_THEN_the_new_ticket_is_number_11()
        {
            TestingTicketDispenser testingTicketDispenser = new TestingTicketDispenser();
            TurnTicket ticket = testingTicketDispenser.GetTurnTicket();
            Assert.AreEqual(11, ticket.TurnNumber);
        }
    }
}

File: TDDMicroExercises.YoursSolutions/Java-viswans/TurnTicketDispenser/src/test/java/tddmicroexercises/turnticketdispenser/TicketDispenserTest.java
package tddmicroexercises.turnticketdispenser;

import junit.framework.Assert;

import org.junit.Test;

public class TicketDispenserTest {
	
	@Test public void
	should_generate_turn_ticket_with_current_number() {
		ITurnNumberSequence seq = new MockTurnNumberSequence(0);
		TicketDispenser td = new TicketDispenser(seq);
		
		Assert.assertEquals(0, td.getTurnTicket().getTurnNumber());
		Assert.assertEquals(1, td.getTurnTicket().getTurnNumber());
	}
	
	@Test public void
	should_able_to_share_turn_number_sequence() {
		ITurnNumberSequence seq = new MockTurnNumberSequence(0);
		TicketDispenser td1 = new TicketDispenser(seq);
		TicketDispenser td2 = new TicketDispenser(seq);
		
		Assert.assertEquals(0, td1.getTurnTicket().getTurnNumber());
		Assert.assertEquals(1, td2.getTurnTicket().getTurnNumber());
	}
	
	static class MockTurnNumberSequence implements ITurnNumberSequence {
		
		private int num;
		
		MockTurnNumberSequence(int num) {
			this.num = num;
		}

		public int getNextTurnNumber() {
			return num++;
		}
		
	}

}
